libavcodec/vvc/ctu.c

```diff
@@ -1806,16 +1886,266 @@ static int inter_data(VVCLocalContext *lc)
+#define PALETTE_RUN_TYPE(c, xc, yc)   PALETTE_VAL(copy_above_indices_flag, c, xc, yc)
+static int palette_subblock_data(VVCLocalContext *lc, int start_comp, int num_comp,
+    const uint8_t *hs, const uint8_t *vs, int max_palette_index, int subset_id, int *prev_run_type,
+    int *prev_run_position, int palette_transpose_flag, int palette_escape_val_present_flag, int *adjust)
+{
+    const CodingUnit    *cu  = lc->cu;
+    const TransformUnit *tu  = cu->tus.head;
+    const VVCSPS        *sps = lc->fc->ps.sps;
+
+    uint8_t run_copy_map[16] = { 0 };
+    uint32_t xc, yc, prev_xc, prev_yc;
+
+    int current_palette_index      = 0;
+    int adjusted_ref_palette_index = 0;
+
+    int min_sub_pos = subset_id << 4;
+    int max_sub_pos = min_sub_pos + 16;
+    int scan_pos;
+
+    int width  = tu->tbs[start_comp].tb_width;
+    int height = tu->tbs[start_comp].tb_height;
+
+    uint32_t (*scan_order)[2] = (uint32_t (*)[2])lc->fc->tab.traverse_scan_order[palette_transpose_flag ? TRAV_VERT : TRAV_HORIZ][av_log2(width) - 1][av_log2(height) - 1];
+
+    max_sub_pos = FFMIN(max_sub_pos, width * height);
+    for (scan_pos = min_sub_pos; scan_pos < max_sub_pos; scan_pos++) {
+        xc = scan_order[scan_pos][0];
+        yc = scan_order[scan_pos][1];
+
+        prev_xc = 0;
+        prev_yc = 0;
+
+        if (scan_pos > 0) {
+            prev_xc = scan_order[scan_pos - 1][0];
+            prev_yc = scan_order[scan_pos - 1][1];
+            if (max_palette_index > 0)
+                run_copy_map[scan_pos - min_sub_pos] = ff_vvc_run_copy_flag(lc, *prev_run_type, *prev_run_position, scan_pos);
+        }
+
+        PALETTE_RUN_TYPE(start_comp, xc, yc) = 0;
+        if (max_palette_index > 0 && !run_copy_map[scan_pos - min_sub_pos]) {
+            if (((!palette_transpose_flag && yc > 0) || (palette_transpose_flag && xc > 0))
+                && PALETTE_RUN_TYPE(start_comp, prev_xc, prev_yc) == 0 && scan_pos > 0) {
+                PALETTE_RUN_TYPE(start_comp, xc, yc) = ff_vvc_copy_above_palette_indices_flag(lc);
+            }
+            *prev_run_type = PALETTE_RUN_TYPE(start_comp, xc, yc);
+            *prev_run_position = scan_pos;
+        } else if (scan_pos > 0) {
+            PALETTE_RUN_TYPE(start_comp, xc, yc) = PALETTE_RUN_TYPE(start_comp, prev_xc, prev_yc);
+        }
+    }
+
+    for (scan_pos = min_sub_pos; scan_pos < max_sub_pos; scan_pos++) {
+        xc = scan_order[scan_pos][0];
+        yc = scan_order[scan_pos][1];
+
+        prev_xc = scan_pos > 0 ? scan_order[scan_pos - 1][0] : 0;
+        prev_yc = scan_pos > 0 ? scan_order[scan_pos - 1][1] : 0;
+
+        current_palette_index = 0;
+        if (max_palette_index > 0 && !run_copy_map[scan_pos - min_sub_pos] && PALETTE_RUN_TYPE(start_comp, xc, yc) == 0) {
+            current_palette_index = ff_vvc_palette_idx_idc(lc, max_palette_index, *adjust);
+            if (scan_pos > 0) {
+                adjusted_ref_palette_index = max_palette_index + 1;
```

Comment:
This assignment will be replaced by the following if-else block. Why doesn't this code follow the specification?

Reply:
You can report a bug to VTM.